### PR TITLE
Add pickle support to DataObject

### PR DESCRIPTION
**Pull request by whophil**
### Overview

Adds simple pickle support to DataObject via `__getstate__` and `__setstate__`.

```python
import pyvista
from pyvista import examples

grid = pyvista.UnstructuredGrid(examples.hexbeamfile)
grid_2 = pickle.loads(pickle.dumps(grid))
```

It also enables DataObjects to be used in parallel workflows, for example, in this workflow which clips a dataset three ways in parallel using `concurrent.futures`. All of the deserialized objects are copies, so this requires N-times memory.

```python
import concurrent.futures as cf
import numpy as np

import pyvista
from pyvista import examples

m = examples.load_airplane()
m.cell_arrays['cell'] = np.ones((m.n_cells, 1))

def clip_mesh(mesh, normal):
    return mesh.clip(normal=normal)

p = cf.ProcessPoolExecutor()
futures = [p.submit(clip_mesh, m, normal) for normal in ('x', 'y', 'z')]
results = [f.result() for f in futures]

p = pyvista.Plotter(shape=(1, 3))
for i in range(3):
    p.subplot(0, i)
    p.add_mesh(results[i])
p.show()

```
image

### Details

The implementation is similar to the one described in https://github.com/pyvista/pyvista-support/issues/59. As @banesullivan , I also encountered the issue when trying to serialize to binary; in this PR the serialization medium is still unfortunately an ASCII string. At least one other person (see https://stackoverflow.com/questions/38747801/) has run into this binary issue, so perhaps this is something which needs to be fixed on the VTK side.

ActiveArrayInfo, formerly a `namedtuple`, is rewritten as a class to support serialization because the FieldAssociation class (stored in attributes of ActiveArrayInfo) uses VTK enums which are not serializable. `namedtuple` cannot be inherited in a normal way, so legacy support for `__getitem__`, `__setitem__`, `__repr__`, `__iter__` is added by means of a `namedtuple` built on-the-fly. Perhaps there's a better way to do this...

### CODE REVIEW

**Comment by JevinJ** [comment]
I would actually prefer to remove ActiveArrayInfo entirely. This means moving active_tensors/vectors/scalars logic to DataSetAttributes where it belongs and would probably be less code overall once moved. The only reason I haven't done it yet is because it would change the API, but it would be much better. 

```python

something.set_active_vectors(name='name', preference='point'):

# becomes
something.point_arrays.set_active_vectors(name=name)
```

**Comment by akaszynski** [comment]
> Somewhat related, I would prefer instead of seeing ActiveArrayInfo being expanding on, remove it entirely. This would mean moving active_tensors/vectors/scalars logic to DataSetAttributes where it belongs and would be less code overall once moved. The only reason I haven't done it yet is because it would change the API.

As for moving forward, I think the best approach would be to merge this PR and then follow up with a PR from @JevinJ that removes `ActiveArrayInfo`.  @JevinJ, you ok with that?

**Comment by JevinJ** [comment]
@akaszynski Yes.

File: tests/test_common.py [hunk 1]
```diff
@@ -981,3 +982,27 @@ def test_cell_bounds(grid):
 def test_cell_type(grid):
     ctype = grid.cell_type(0)
     assert isinstance(ctype, int)
+
+
+def test_serialize_deserialize(grid):
```
Comment (akaszynski) [inline tests/test_common.py:987]: I'd like to see a parameterized test that serializes and deserializes a bunch of pyvista objects.  Ideally, the big five:
- `PolyData`
- `UnstructuredGrid`
- `UniformGrid`
- `StrucutredGrid`
- `RectilinearGrid`

For reference, see:
https://github.com/pyvista/pyvista/blob/87a2003ad828b0b8773e95caf7ada43bb315d447/tests/test_filters.py#L11-L17

**Review by akaszynski — APPROVED** [review]
Approved with comments.  I'd like to see additional tests to ensure we're supporting the "big five".  Otherwise, nice work!  Would be great if we could get around using ASCII, but being able to pickle an object is better than not being able to.  Probably should open an issue on VTK regarding this.

Once this is merged we can work on cleaning up `ActiveArrayInfo`.  @JevinJ, you're lightning fast in your responses.

File: tests/test_common.py (hunk 1 above)
Comment (whophil) [reply tests/test_common.py]: Thanks, that is a useful snippet. This is implemented now.

**Pull request merged**